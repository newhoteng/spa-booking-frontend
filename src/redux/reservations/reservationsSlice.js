import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const postUrl = 'http://localhost:3001/api/v1/reservations';

const initialState = {
  userReservations: [],
  isLoading: false,
  error: undefined,
};

export const getUserReservations = createAsyncThunk(
  'reservations/getUserReservations',
  async () => {
    try {
      const resp = await axios.get(`http://localhost:3001/api/v1/users/${JSON.parse(localStorage.getItem('user')).id}/reservations`);
      return resp.data;
    } catch (error) {
      return (error.message);
    }
  },
);

// Add a new service
export const postReservation = createAsyncThunk(
  'reservations/postReservation',
  async (reservationData) => {
    const response = await axios.post(postUrl, reservationData);
    return response.data;
  },
);

const reservationsSlice = createSlice({
  name: 'userReservations',
  initialState,
  reducers: {

  },
  extraReducers: (builder) => {
    // getUserReservations
    builder.addCase(getUserReservations.pending, (state) => {
      state.isLoading = true;
    });
    builder.addCase(getUserReservations.fulfilled, (state, action) => {
      state.isLoading = false;
      state.error = false;
      state.userReservations = action.payload;
    });
    builder.addCase(getUserReservations.rejected, (state, action) => {
      state.isLoading = false;
      state.error = action.payload;
    });
    // postReservation
    builder.addCase(postReservation.pending, (state) => {
      state.isLoading = true;
    });
    builder.addCase(postReservation.fulfilled, (state, action) => {
      state.isLoading = false;
      state.userReservations.push(action.payload);
    });
    builder.addCase(postReservation.rejected, (state, action) => {
      state.error = action.payload;
      state.isLoading = false;
    });
  },
});

export const { addReservation } = reservationsSlice.actions;
export default reservationsSlice.reducer;
