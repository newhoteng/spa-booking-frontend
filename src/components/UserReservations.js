import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { getUserReservations } from '../redux/reservations/reservationsSlice';
import { fetchAllServices } from '../redux/serviceSlice';
import styles from '../styles/Forms.module.css';

function UserReservations() {
  // Use Redux selectors to access data from the store
  const { userReservations, isLoading, error } = useSelector((store) => store.userReservations);
  const { services } = useSelector((store) => store.services);
  const dispatch = useDispatch();

  // useEffect hook to dispatch actions when the component mounts
  useEffect(() => {
    dispatch(fetchAllServices());
    dispatch(getUserReservations());
  }, [dispatch]);

  // Render a loading message while data is being fetched
  if (isLoading) {
    return (
      <ul className={`${styles.reserveForm}`}>
        <h1>Loading...</h1>
      </ul>
    );
  }
  // Render an error message if there's an error during data retrieval
  if (error) {
    return (
      <ul className={`${styles.reserveForm}`}>
        <h1>Something went wrong</h1>
      </ul>
    );
  }

  return (
    <div className={`${styles.reserveForm} ${styles.flexColumn}`}>
      {!isLoading && userReservations.length === 0 && (
        <p className={`${styles.headline}`}><i>You have no reservations.</i></p>
      )}
      {userReservations.length !== 0 && (
        <p className={`${styles.headline}`}>
          <i>
            You have
            {' '}
            {userReservations.length}
            {' '}
            reservation(s)
          </i>
        </p>
      )}
      {userReservations.length !== 0 && (
        <table>
          <thead>
            <tr>
              <th>Service</th>
              <th>Date</th>
              <th>City</th>
            </tr>
          </thead>
          <tbody>
            {userReservations?.map((rs) => (
              <tr key={rs.id}>
                <td>
                  {services.find((service) => service.id === rs.spa_service_id)?.name.toUpperCase()}
                </td>
                <td>{rs.date}</td>
                <td>{rs.city}</td>
              </tr>
            ))}
          </tbody>
        </table>
      )}
    </div>
  );
}

export default UserReservations;
