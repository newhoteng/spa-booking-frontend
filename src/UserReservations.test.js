import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
// import { store } from './redux/store';
import { getUserReservations, postReservation } from './redux/reservations/reservationsSlice';

// Component imports
import UserReservations from './components/UserReservations';

const mockStore = configureStore([thunk]);

describe('UserReservations component', () => {
  test('fetches users reservations correctly', async () => {
    const store = mockStore({
      userReservations: {
        userReservations: [],
      },
      services: {
        services: [],
      },
    });

    // Dispatch the asynchronous action
    await store.dispatch(getUserReservations());

    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <UserReservations />
        </BrowserRouter>
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });

  test('adds users reservations correctly', async () => {
    const store = mockStore({
      userReservations: {
        userReservations: [],
      },
      services: {
        services: [],
      },
    });

    // Dispatch the asynchronous action
    await store.dispatch(postReservation());

    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <UserReservations />
        </BrowserRouter>
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });
});

it('userReservations page renders correctly', () => {
  const store = mockStore({
    userReservations: {
      userReservations: [],
    },
    services: {
      services: [],
    },
  });

  const userReservations = renderer
    .create(
      <Provider store={store}>
        <UserReservations page="/reservations">Reservations Page</UserReservations>
      </Provider>,
    )
    .toJSON();
  expect(userReservations).toMatchSnapshot();
});

it('renders Current component', () => {
  const store = mockStore({
    userReservations: {
      userReservations: [],
    },
    services: {
      services: [],
    },
  });

  render(
    <Provider store={store}>
      <UserReservations />
    </Provider>,
  );
});
