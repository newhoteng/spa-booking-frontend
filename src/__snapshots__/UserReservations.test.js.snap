// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserReservations component adds users reservations correctly 1`] = `
<div>
  <div
    class="reserveForm flexColumn"
  >
    <p
      class="headline"
    >
      <i>
        You have no reservations.
      </i>
    </p>
  </div>
</div>
`;

exports[`UserReservations component fetches users reservations correctly 1`] = `
<div>
  <div
    class="reserveForm flexColumn"
  >
    <p
      class="headline"
    >
      <i>
        You have no reservations.
      </i>
    </p>
  </div>
</div>
`;

exports[`userReservations page renders correctly 1`] = `
<div
  className="reserveForm flexColumn"
>
  <p
    className="headline"
  >
    <i>
      You have no reservations.
    </i>
  </p>
</div>
`;
